src/blocks/CalligraphyApp.tsx

```diff
@@ -0,0 +1,753 @@
+                );
+              }
+            }
+          }
+        }
+        
+
+      s.image(buffer,0,0)
+      //if the clear switch has been toggled by the toolbar, we save an undo frame, clear, and align the local tracking variable to match it so we cathc the next flip
+      if (state.current.canvasClearSwitch != localBufferClearSwitch) {
+        saveUndoFrame();
+        buffer.clear();
+        s.clear();
+        s.image(buffer,0,0)
+        saveDataToSeamModel();
+        localBufferClearSwitch = state.current.canvasClearSwitch
+      }
+      //If the undo switch is tripped, clear the buffer, image the undo buffer (should be one stroke behind) and remove that undo frame from the stack
+      if (state.current.canvasUndoSwitch != localUndoSwitch) {
+        undo();
+        localUndoSwitch = state.current.canvasUndoSwitch
+      }
+    }
+    
+    /**Our click/touch handlers check if mouse is in canvas and ignore it if not */
+    s.mousePressed = () => {
+      if (mouseOffCanvas()) return;
+      onStrokeStart()
+    };
+    s.touchStarted = () => {
+      if (mouseOffCanvas()) return;
+      onStrokeStart()
+    }
+    s.touchEnded = () => {
+      inStroke = false;
+      saveDataToSeamModel();
+    }
+    s.mouseReleased = () => {
+      inStroke = false;
+      saveDataToSeamModel();
+    }
+    const onStrokeStart = () => {
+      /**Set the stroking status to true (so draw() knows to draw; this decouples drawing state from All mouseIsPressed() situations) */
+      inStroke = true;
+      /** Whenever we start a drawing move, first save the pre-stroke state as a frame in the undo stack */
+      saveUndoFrame();
+
+      /** Reset the brush position/velocity and total stroke lenghth */
+      ink.currentStrokeTotalLength = 0;
+      ink.previousStrokeWidth = 0;
+      [ink.brushX, ink.brushY] = [s.mouseX, s.mouseY];
+      [spray.brushX, spray.brushY] = [s.mouseX, s.mouseY];
+      spray.prevSpeed = 0;
+      [streak.brushX, streak.brushY] = [s.mouseX, s.mouseY];
+      [lines.brushX, lines.brushY] = [s.mouseX, s.mouseY];
+      lines.lineSpacingOffsets = Array.from({length: lines.LINES}, ()=>s.random(-lines.lineSpacingVar/2,lines.lineSpacingVar/2))
+
+    }
+
+    const undo = () => {
+      if (undoBufferStack.length === 0) return;
+      buffer.clear();
+      buffer.image(undoBufferStack.pop()!,0,0)
+      s.image(buffer,0,0)
+      saveDataToSeamModel();
+    }
+    
+    const saveUndoFrame = () => {
+      buffer.loadPixels(); //load the pixels array for the buffer
+      const newUndoBufferFrame = buffer.get() //get it as a p5.Image object
+      undoBufferStack.push(newUndoBufferFrame) //push it onto the undo state stack
+      if (undoBufferStack.length > 150) undoBufferStack.splice(0,undoBufferStack.length - 150)
+    }
+    const saveDataToSeamModel = () => {
+      const canvas:HTMLCanvasElement = s.drawingContext.canvas
+      const imageData = canvas.toDataURL();
```

Comment:
for some future consideration (not needed to do in this PR): we could consider using `image/webp` as a format here for smaller image sizes https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL